### PR TITLE
Add a game fix to set EE Cyclerate at 100% while FMVs are playing

### PR DESCRIPTION
**Pull request by Siphonay**
This fixes FMVs being slowed down when EE Cyclerate is set at high values

(Edit on 05/12)
This fixes FMV cutscenes being played at a slower speed than usual when the EE cyclerate is set to high values on some games, notably the Ratchet & Clank series.

Another option would be to merge this functionality in the "FMV in Software Mode" fix and rename it to "FMV Safe Mode". I opened a pull request for that: #2435

### CODE REVIEW

**Comment by pgert** [comment]
Can you give examples of games that makes use of this?
Can the hack be applied through GameIndex, and if so please add info about the hack in the Game Fixes info section.
"Set EE Cycle rate to 100% when a FMV plays" ought to be "Set the EE Cyclerate to 100% when an FMV plays".

**Comment by refractionpcsx2** [comment]
I would kind of like to see this hack being merged with the software mode FMV hack so we don't have yet ANOTHER gamefix :P Both are related to FMV's and I don't see any harm in doing both just for them IMO.

Of course that depends what other people think..

**Comment by Siphonay** [comment]
@pgert I've experienced this with the Ratchet & Clank and the Sly games. I guess this applies to all MPEG cutscenes but I haven't tested more games to back this claim up.

@refractionpcsx2 this is something I considered, but I haven't been able to figure out what would be best out of the two options, so I decided to put this pull request as is and see people's feedback, so thanks!
I also considered making this a speedhack rather than a gamefix, I feel like it would make more sense if how much cutscenes were slowed down was related with how powerful the machines the emulator is running on was, but since I only have two computers with similarly powerful CPUs, I obtained pretty much the same results on both so I can't say this is the case

**Comment by refractionpcsx2** [comment]
Hmm no it would make less sense as a speedhack as you aren't increasing the speed in any way :P But I guess a gamefix on its own wouldn't make much sense either as you'd need to have the speedhack enabled in the first place.  That's a tough one.

I guess that's one reason to have a merged gamefix :P

**Comment by Siphonay** [comment]
I can get behind the idea of a global FMV safe mode gamefix. Should I wait for more feedback before doing that?

**Comment by refractionpcsx2** [comment]
Yes, I'm going to annoyingly tag some people to get some more feedback on the idea

@gregory38 @ssakash @KrossX @avih @sudonim1 @ramapcsx2 @lightningterror @MrCK1 @turtleli @FlatOutPS2 @gigaherz 

My dudes, what do you think? See my first comment in this thread.

**Comment by MrCK1** [comment]
I don't want another gamefix, but I'd be fine if a compromise was found to merge it with an existing hack, ect.

**Comment by Siphonay** [comment]
@MrCK1 
Well maybe this is something I should have detailed in my initial post but I've done full playthrough of Ratchet & Clank games on PCSX2, and the games all ran better with the EE Cyclerate all the way up to 300%, but this had the effect to slow down FMV cutscenes to around 80%, which made them pretty unenjoyable, so while they were playing I had to minimize the emulator window, and go reset the EE cyclerate to 100%. This fix is a way to automatize this process.

**Comment by ssakash** [comment]
I like ref's idea of integrating it into a "FMV safe mode", seems more cleaner than adding various scenario specific hacks.

**Comment by gregory38** [comment]
Could this new hack be always enabled ? Or is it unsafe ?

**Comment by Siphonay** [comment]
If you mean unsafe as "emulation instability", I don't think setting the EE cyclerate to the default value is unsafe as it is the setting that (according to the UI) has the least compatibility problems.

It term of bugs, as of now: if you stop the emulation while a cutscene is playing, the cyclerate won't reset to what it was before the cutscene started, and if you change the cyclerate while a cutscene is playing, it will be reset to the one that was in effect before the cutscene started.

**Comment by mirh** [comment]
> The we can call it something like "FMV Safe Mode" or something.

That's a smart idea. 
Though, correct me if I'm wrong, there's no "hard reason" for hardware mode to create problems with FMVs (and so that's kinda of a workaround, just like all other hacks AFAIU)
Instead touching timing.. I wonder if there are even FMVs that wouldn't have a different speed?

**Comment by lightningterror** [comment]
Yes I like ref's idea as well, merge the two hacks in to one.

+1 for FMV Safe Mode

File: pcsx2/Config.h [hunk 1]
```diff
@@ -326,7 +327,7 @@ struct Pcsx2Config
 
 				OpEqu( LimitScalar )			&&
 				OpEqu( FramerateNTSC )			&&
-				OpEqu( FrameratePAL )			&&
+		 		OpEqu( FrameratePAL )			&&
```
Comment (lightningterror) [inline pcsx2/Config.h]: Please undo this change, the whitespace was fine.

File: pcsx2/Pcsx2Config.cpp [hunk 2]
```diff
@@ -331,6 +332,7 @@ void Pcsx2Config::GamefixOptions::Set( GamefixId id, bool enabled )
 		case Fix_FMVinSoftware:	FMVinSoftwareHack	= enabled;  break;
 		case Fix_GoemonTlbMiss: GoemonTlbHack		= enabled;  break;
 		case Fix_ScarfaceIbit:  ScarfaceIbit        = enabled;  break;
+		case Fix_FMVEECycle:    FMVEECycleHack      = enabled; break;
```
Comment (lightningterror) [inline pcsx2/Pcsx2Config.cpp]: Add an extra space here, it looks out of place compared to the rest above.

**Comment by RedPanda4552** [comment]
+1 for merge. My issue with having it always on, is there may be some (although we're probably scraping the edge case barrel here) situations where someone has to play a game with the EE cyclerate anywhere from level -1 to -3 to hold full speed, and it kicking up to level 0 on a FMV may actually slow it down again, having the inverse effect as the hack intends. At least, if the trend of "More EE cycles = less FMV FPS" holds up.

**Comment by refractionpcsx2** [comment]
I don't know if any cases that need a - ee cycle rate to hold full fps on a video, the ps2 was more than capable of video playback :p

**Comment by lightningterror** [comment]
Well we can also change the gui option to a drop down list how vsync works, if there is an issue if the two are merged.

FMV mode: Switch to SW mode during FMV, Switch to SW + Cycle Rate, 100 Cyclerate.

File: pcsx2/gui/Panels/GameFixesPanel.cpp [hunk 3]
```diff
@@ -107,6 +107,10 @@ Panels::GameFixesPanel::GameFixesPanel( wxWindow* parent )
 		{
 			_("VU I bit Hack avoid constant recompilation (Scarface The World Is Yours)"),
 			wxEmptyString
+		},
+		{
+			_("Set EE Cycle rate to 100% when a FMV plays"),
```
Comment (MrCK1) [inline pcsx2/gui/Panels/GameFixesPanel.cpp]: Clarification should be added here that the EE cyclerate will reset to its normal speed (100% isn't very clear)

The notable cases where it is needed should be noted as well, such as R&C.

**Comment by KrossX** [comment]
Yet another +1 for FMV Safe Mode. Also, thread gets updated and people stay notified. =P

**Comment by turtleli** [comment]
> I would kind of like to see this hack being merged with the software mode FMV hack so we don't have yet ANOTHER gamefix :P Both are related to FMV's and I don't see any harm in doing both just for them IMO. The we can call it something like "FMV Safe Mode" or something.

Well, the "software mode" hack switches renderer regardless of whether the current renderer is actually software or hardware. The behaviour of that hack might need to be changed first before merging the 100% cyclerate hack with the "software mode" hack together into an FMV safe mode hack.

> Could this new hack be always enabled ?

Might not be a good idea - see #1634 (Valkyrie Profile 2 PAL FMV locks up at the end unless EE cyclerate is -1 or lower).

**Comment by Siphonay** [comment]
Since the consensus seems to be merging this with the Software FMV hack I think it would be better to close this pull request and do a new one (while also correcting the behavior of the renderer switch if I can figure out how to do that)

What do you think? (I'll try not to mess up indentation this time)

**Comment by lightningterror** [comment]
Well if you want to start over sure, if it were me tho I'd drop the commits during rebase so I would be starting clean anyway without closing and creating a new PR.

**Comment by refractionpcsx2** [comment]
> Well, the "software mode" hack switches renderer regardless of whether the current renderer is actually software or hardware. The behaviour of that hack might need to be changed first before merging the 100% cyclerate hack with the "software mode" hack together into an FMV safe mode hack.

TBH if you're already in software mode, you shouldn't be using the hack anyway, so there would currently be an issue with it being on.  So IMO that's outside the scope of this PR.

**Comment by gregory38** [comment]
I hate to be that guy. But I don't think merging GSdx workaround + this EE timing is a good idea. In the end, it will always be enabled by user, for 1 issue or another.

Except few complex corner case, GSdx should be improved.

**Comment by turtleli** [comment]
> TBH if you're already in software mode, you shouldn't be using the hack anyway, so there would currently be an issue with it being on. So IMO that's outside the scope of this PR, but should probably be looked in to regardless of the result of this.

The hack is in the GameDB for Silent Hill 2 and 3, so it's automatically switched on for any Software mode users playing those games. But yeah, the hack behaviour might be out of scope for this PR.

**Comment by willkuer** [comment]
I'd suggest to enable it by default and disable it via ini setting. Accordingly no merge.

**Comment by refractionpcsx2** [comment]
> Is there any reason to keep this open now? We do have #2435 which is an updated version of this PR unless I'm wrong.

Technically there is no reason, but suddenly people seem to be against the idea in the other version for no apparent reason :P

**Comment by Siphonay** [comment]
@ssakash this one adds a separate gamefix and the other one merges it with an existing one, I kept both open so the PCSX2 team can decide which one to merge (if they decide to merge any)

**Comment by ssakash** [comment]
To be honest, I prefer the other version more than this one. This doesn't seem like something worthy enough to be added as a separate gamefix and we do have lots of gamefixes now. :/

**Comment by refractionpcsx2** [comment]
I completely agree with that.

**Comment by Siphonay** [comment]
The consensus seems to be that we're going with the other PR, so I'm closing this one. Thank you all for your consideration! :)
